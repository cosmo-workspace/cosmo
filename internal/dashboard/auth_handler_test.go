package dashboard

import (
	"context"
	"net/http"

	. "github.com/cosmo-workspace/cosmo/pkg/snap"
	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"

	"github.com/bufbuild/connect-go"
	emptypb "google.golang.org/protobuf/types/known/emptypb"
	"google.golang.org/protobuf/types/known/timestamppb"

	cosmov1alpha1 "github.com/cosmo-workspace/cosmo/api/v1alpha1"
	dashv1alpha1 "github.com/cosmo-workspace/cosmo/proto/gen/dashboard/v1alpha1"
	"github.com/cosmo-workspace/cosmo/proto/gen/dashboard/v1alpha1/dashboardv1alpha1connect"
)

var _ = Describe("Dashboard server [auth]", func() {

	var client dashboardv1alpha1connect.AuthServiceClient

	BeforeEach(func() {
		testUtil.CreateLoginUser("normal-user", "user", nil, cosmov1alpha1.UserAuthTypePasswordSecert, "password1")
		testUtil.CreateLoginUser("admin-user", "admin", []cosmov1alpha1.UserRole{cosmov1alpha1.PrivilegedRole}, cosmov1alpha1.UserAuthTypePasswordSecert, "password2")
		testUtil.CreateLoginUser("ldap-user", "ldap-u", []cosmov1alpha1.UserRole{cosmov1alpha1.PrivilegedRole}, cosmov1alpha1.UserAuthTypeLDAP, "")
		client = dashboardv1alpha1connect.NewAuthServiceClient(http.DefaultClient, "http://localhost:8888")
	})

	AfterEach(func() {
		clientMock.Clear()
		testUtil.DeleteCosmoUserAll()
	})

	//==================================================================================

	//==================================================================================
	Describe("[Login]", func() {

		run_test := func(req *dashv1alpha1.LoginRequest) {
			By("---------------test start----------------")
			ctx := context.Background()
			res, err := client.Login(ctx, connect.NewRequest(req))
			if err == nil {
				Expect(res.Msg.ExpireAt).ShouldNot(BeNil())
				res.Msg.ExpireAt = &timestamppb.Timestamp{}
				Ω(res.Msg).To(MatchSnapShot())
				Expect(res.Header().Get("Set-Cookie")).ShouldNot(BeNil())
			} else {
				Ω(err.Error()).To(MatchSnapShot())
				Expect(res).Should(BeNil())
			}
			By("---------------test end---------------")
		}

		DescribeTable("✅ success in normal context:",
			run_test,
			Entry(nil, &dashv1alpha1.LoginRequest{UserName: "normal-user", Password: "password1"}),
			Entry(nil, &dashv1alpha1.LoginRequest{UserName: "admin-user", Password: "password2"}),
			Entry(nil, &dashv1alpha1.LoginRequest{UserName: "ldap-user", Password: "password"}),
		)

		DescribeTable("❌ fail with invalid request:",
			run_test,
			Entry(nil, &dashv1alpha1.LoginRequest{Password: "password1"}),
			Entry(nil, &dashv1alpha1.LoginRequest{UserName: "normal-user"}),
			Entry(nil, &dashv1alpha1.LoginRequest{UserName: "normal-user", Password: "invalid"}),
			Entry(nil, &dashv1alpha1.LoginRequest{UserName: "ldap-user", Password: "invalid"}),
			Entry(nil, &dashv1alpha1.LoginRequest{UserName: "xxxxxxx", Password: "password1"}),
		)
	})

	//==================================================================================
	Describe("[Verify]", func() {

		run_test := func(sessionType string) {
			ctx := context.Background()
			var session string
			switch sessionType {
			case "nil session":
			case "logined session":
				session = test_Login("normal-user", "password1")
			case "logouted session":
				session = test_Login("normal-user", "password1")
				logoutResp, _ := client.Logout(ctx, NewRequestWithSession(&emptypb.Empty{}, session))
				session = logoutResp.Header().Get("Set-Cookie")
			}
			By("---------------test start----------------")
			res, err := client.Verify(ctx, NewRequestWithSession(&emptypb.Empty{}, session))
			if err == nil {
				Expect(res.Msg.ExpireAt).ShouldNot(BeNil())
				res.Msg.ExpireAt = &timestamppb.Timestamp{}
				Ω(res.Msg).To(MatchSnapShot())
				Expect(res.Header().Get("Set-Cookie")).Should(BeEmpty())
			} else {
				Ω(err.Error()).To(MatchSnapShot())
				Expect(res).Should(BeNil())
			}
			By("---------------test end---------------")
		}

		DescribeTable("✅ success in normal context:",
			run_test,
			Entry(nil, "logined session"),
		)

		DescribeTable("❌ fail with invalid request:",
			run_test,
			Entry(nil, "logouted session"),
			Entry(nil, "nil session"),
		)
	})

	//==================================================================================
	Describe("[Logout]", func() {

		run_test := func(sessionType string) {
			ctx := context.Background()
			var session string
			switch sessionType {
			case "nil session":
			case "logined session":
				session = test_Login("normal-user", "password1")
			case "logouted session":
				session = test_Login("normal-user", "password1")
				logoutResp, _ := client.Logout(ctx, NewRequestWithSession(&emptypb.Empty{}, session))
				session = logoutResp.Header().Get("Set-Cookie")
			}
			By("---------------test start----------------")
			res, err := client.Logout(ctx, NewRequestWithSession(&emptypb.Empty{}, session))
			if err == nil {
				Ω(res.Msg).To(MatchSnapShot())
				Ω(res.Header().Get("Set-Cookie")).To(MatchSnapShot())
			} else {
				Ω(err.Error()).To(MatchSnapShot())
				Expect(res).Should(BeNil())
			}
			By("---------------test end---------------")
		}

		DescribeTable("✅ success in normal context:",
			run_test,
			Entry(nil, "logined session"),
		)

		DescribeTable("❌ fail with invalid request:",
			run_test,
			Entry(nil, "logouted session"),
			Entry(nil, "nil session"),
		)
	})
})
