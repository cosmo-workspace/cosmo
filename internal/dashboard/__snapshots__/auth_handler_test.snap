['Dashboard server [auth] [Login] ✅ success in normal context: Entry: user_name:"admin-user" password:"password2" 1']
SnapShot = """
{
  \"user_name\": \"admin-user\",
  \"expire_at\": {}
}
"""

['Dashboard server [auth] [Login] ✅ success in normal context: Entry: user_name:"ldap-user" password:"password" 1']
SnapShot = """
{
  \"user_name\": \"ldap-user\",
  \"expire_at\": {}
}
"""

['Dashboard server [auth] [Login] ✅ success in normal context: Entry: user_name:"normal-user" password:"password1" 1']
SnapShot = """
{
  \"user_name\": \"normal-user\",
  \"expire_at\": {}
}
"""

['Dashboard server [auth] [Login] ❌ fail with invalid request: Entry: password:"password1" 1']
SnapShot = 'permission_denied: incorrect user or password'

['Dashboard server [auth] [Login] ❌ fail with invalid request: Entry: user_name:"ldap-user" password:"invalid" 1']
SnapShot = 'permission_denied: incorrect user or password'

['Dashboard server [auth] [Login] ❌ fail with invalid request: Entry: user_name:"normal-user" 1']
SnapShot = 'permission_denied: incorrect user or password'

['Dashboard server [auth] [Login] ❌ fail with invalid request: Entry: user_name:"normal-user" password:"invalid" 1']
SnapShot = 'permission_denied: incorrect user or password'

['Dashboard server [auth] [Login] ❌ fail with invalid request: Entry: user_name:"xxxxxxx" password:"password1" 1']
SnapShot = 'permission_denied: incorrect user or password'

['Dashboard server [auth] [Logout] ✅ success in normal context: Entry: logined session 1']
SnapShot = """
{}
"""

['Dashboard server [auth] [Logout] ✅ success in normal context: Entry: logined session 2']
SnapShot = 'test-server=; Path=/; Domain=test.domain; Max-Age=0; HttpOnly; SameSite=Lax'

['Dashboard server [auth] [Logout] ❌ fail with invalid request: Entry: logouted session 1']
SnapShot = 'unauthenticated: failed to get session from store'

['Dashboard server [auth] [Logout] ❌ fail with invalid request: Entry: nil session 1']
SnapShot = 'unauthenticated: session is not found'

['Dashboard server [auth] [Verify] ✅ success in normal context: Entry: logined session 1']
SnapShot = """
{
  \"user_name\": \"normal-user\",
  \"expire_at\": {}
}
"""

['Dashboard server [auth] [Verify] ❌ fail with invalid request: Entry: logouted session 1']
SnapShot = 'unauthenticated: failed to get session from store'

['Dashboard server [auth] [Verify] ❌ fail with invalid request: Entry: nil session 1']
SnapShot = 'unauthenticated: session is not found'
